python/tritonserver/_api/_datautils.py

```diff
@@ -0,0 +1,753 @@
+    def memory_type_id(self) -> int:
+        return self.memory_buffer.memory_type_id
+
+    @property
+    def size(self) -> int:
+        return self.memory_buffer.size
+
+    def __dlpack__(self, *, stream=None):
+        #        if not (stream is None or (isinstance(stream, int) and stream == 0)):
+        #           raise UnsupportedError(
+        #              f"DLPack stream synchronization on {stream} not currently supported"
+        #         )
+
+        dl_managed_tensor = Tensor._create_managed_tensor()
+        dl_managed_tensor.dl_tensor.data = self.data_ptr
+        dl_managed_tensor.dl_tensor.device = _dlpack.DLDevice(
+            TRITON_MEMORY_TYPE_TO_DLPACK_DEVICE_TYPE[self.memory_type],
+            self.memory_type_id,
+        )
+
+        dl_managed_tensor.dl_tensor.dtype = TRITON_TO_DLPACK_DTYPE[self.data_type]
+        dl_managed_tensor.dl_tensor.ndim = len(self.shape)
+        dl_managed_tensor.dl_tensor.shape = (ctypes.c_int64 * len(self.shape))(
+            *self.shape
+        )
+        dl_managed_tensor.dl_tensor.strides = ctypes.POINTER(ctypes.c_int64)()
+        dl_managed_tensor.dl_tensor.byte_offset = 0
+        dl_managed_tensor.deleter = Tensor._managed_tensor_deleter
+
+        self._set_dlpack_manager_ctx(dl_managed_tensor)
+        pycapsule = ctypes.pythonapi.PyCapsule_New(
+            ctypes.byref(dl_managed_tensor),
+            _dlpack.c_str_dltensor,
+            Tensor._pycapsule_deleter,
+        )
+        return pycapsule
+
+    def __dlpack_device__(self):
+        return (
+            TRITON_MEMORY_TYPE_TO_DLPACK_DEVICE_TYPE[self.memory_type],
+            self.memory_type_id,
+        )
+
+    def to_bytes_array(self) -> numpy.ndarray:
+        if self.data_type != DataType.BYTES:
+            raise InvalidArgumentError(
+                f"Tensor has data type {self.data_type} not {DataType.BYTES}"
+            )
+
+        # Reshape into 1d array of bytes on host
+        original_data_type = self.data_type
+        original_shape = self.shape
+        self.data_type = DataType.UINT8
+        self.shape = [self.size]
+        numpy_ndarray = self._to_numpy_on_host()
+
+        # Deserialize bytes array and reshape
+        self.shape = original_shape
+        self.data_type = original_data_type
+        return Tensor._deserialize_bytes_array(numpy_ndarray).reshape(self.shape)
+
+    @staticmethod
+    def from_memory_buffer(data_type, shape, memory_buffer):
+        return Tensor(data_type, shape, memory_buffer)
+
+    @staticmethod
+    def from_object(obj: Any):
+        if type(obj) in Tensor._from_converters:
+            return Tensor._from_converters[type(obj)](obj)
+        elif hasattr(obj, "__dlpack__"):
+            return Tensor.from_dlpack(obj)
+        else:
+            raise InvalidArgumentError(
+                f"Input type {type(obj)} not supported. Must be one of {list(Tensor._from_converters.keys())} or the type must support __dlpack__"
+            )
```

Comment:
I still think just having `from_dlpack` is sufficient... happy to discuss more offline.